File: src/main/java/toast/pine/SocialSystem/SocialManager.java
package toast.pine.SocialSystem;

import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.player.AsyncPlayerChatEvent;
import toast.pine.Classes.Quests.Quest;

import java.util.*;
import java.util.function.Consumer;

public class SocialManager implements Listener {

    private final Set<Group> groups = new HashSet<>();
    private final Map<UUID, Consumer<String>> nextMessageCallbacks = new HashMap<>();
    private final Set<Quest> questSet = new HashSet<>();


    public Map<UUID, Consumer<String>> getNextMessageCallbacks() {
        return Map.copyOf(nextMessageCallbacks);
    }


    public void addNextMessageCallback(UUID playerId, Consumer<String> callback) {
        nextMessageCallbacks.put(playerId, callback);
    }

    public void removeNextMessageCallback(UUID playerId) {
        nextMessageCallbacks.remove(playerId);
    }

    /**
     * Gets a group by String name
     * @param groupName The name of the group.
     * @return The group.
     */
    public Group findGroupByName(String groupName) {
        for (Group group : groups) {
            if (group.groupName.equals(groupName)) {
                return group;
            }
        }
        return null;
    }


    void addGroup(Group group) {
        groups.add(group);
    }


    void deleteGroup(Group group) {
        group.clearGroup();
        groups.remove(group);
    }

    public void addQuest(Quest quest) {
        questSet.add(quest);
    }

    public void removeQuest(Quest quest) {
        questSet.remove(quest);
    }

    public Set<Quest> getQuests() {
        return Set.copyOf(questSet);
    }



    @EventHandler
    private void onChat(AsyncPlayerChatEvent event) {
        Player receiver = event.getPlayer();
        UUID playerId = receiver.getUniqueId();

        if (nextMessageCallbacks.containsKey(playerId)) {
            String message = event.getMessage();

            Consumer<String> callback = nextMessageCallbacks.remove(playerId);
            callback.accept(message);
        }
    }

}

File: src/main/java/pine/toast/socialSystem/SocialManager.kt
package toast.pine.socialSystem

import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.player.AsyncPlayerChatEvent
import toast.pine.Classes.Quests.Quest
import java.util.*
import java.util.function.Consumer

class SocialManager : Listener {
    private val groups: MutableSet<Group> = HashSet()
    private val nextMessageCallbacks: MutableMap<UUID, Consumer<String>?> = HashMap<UUID, Consumer<String>?>()
    private val questSet: MutableSet<Quest> = HashSet<Quest>()

    fun getNextMessageCallbacks(): Map<UUID, Consumer<String>?> {
        return java.util.Map.copyOf<UUID, Consumer<String>>(nextMessageCallbacks)
    }

    fun addNextMessageCallback(playerId: UUID, callback: Consumer<String>?) {
        nextMessageCallbacks[playerId] = callback
    }

    fun removeNextMessageCallback(playerId: UUID) {
        nextMessageCallbacks.remove(playerId)
    }

    /**
     * Gets a group by String name
     * @param groupName The name of the group.
     * @return The group.
     */
    fun findGroupByName(groupName: String): Group? {
        for (group in groups) {
            if (group.groupName == groupName) {
                return group
            }
        }
        return null
    }

    fun addGroup(group: Group) {
        groups.add(group)
    }

    fun deleteGroup(group: Group) {
        group.clearGroup()
        groups.remove(group)
    }

    fun addQuest(quest: Quest) {
        questSet.add(quest)
    }

    fun removeQuest(quest: Quest) {
        questSet.remove(quest)
    }

    val quests: Set<Any>
        get() = java.util.Set.copyOf(questSet)


    @EventHandler
    private fun onChat(event: AsyncPlayerChatEvent) {
        val receiver: Player = event.getPlayer()
        val playerId: UUID = receiver.getUniqueId()
        if (nextMessageCallbacks.containsKey(playerId)) {
            val message: String = event.getMessage()
            val callback = nextMessageCallbacks.remove(playerId)!!
            callback.accept(message)
        }
    }
}
